### DISCUSSION TITLE
Debugging Help for a new-ish nix user

### DISCUSSION THREAD

**Discussion by bassforce86** [nix-community/nixvim#2225] [Q&A] [answered]
I've created a standalone nixvim config, available [here](https://gitlab.com/bf86/flakes/neovim) but i have an odd problem.

When running this against `nix flake check .` and `nix run .#nvim` it works as i'd expect (dashboard is loaded with my shortcuts) 

But when I include this as a home-manager package in a home-manager flake, I get no start-up dashboard, just takes me straight to NetRW.

(home manager inclusion method is in the readme).

Has anyone else seen this behaviour from a standalone installation?

**Comment by MattSturgeon** [comment]
Are you installing the home-manager _module_ or are you including the package you build in your standalone flake in `home.packages`?

See [Standalone Usage](https://nix-community.github.io/nixvim/modules/standalone.html) in the docs.

Sidenote: a standalone package doesn't _need_ to also be in a standalone flake, since one flake can have many outputs.

**Reply by bassforce86** [reply]
Ah sorry.

For clarification I am using the standalone output in my home.packages.

**Reply by MattSturgeon** [reply]
Yes, the issue is here: [flake.nix L61](https://gitlab.com/bf86/flakes/home-manager/-/blob/649a5e403df86a772593fd30f597eb4cbe417b2a/flake.nix#L61). You are importing our home-manager module.

Instead import a module along the lines of:

```nix
{ home.packages = [ inputs.neovim.packages.${system}.default ]; }
```

**Reply by MattSturgeon** [reply]
Ah, I see you do that already [here](https://gitlab.com/bf86/flakes/home-manager/-/blob/649a5e403df86a772593fd30f597eb4cbe417b2a/home.nix#L37).

Maybe the fact you have the standalone package _and_ the hm module is conflicting.

**Reply by bassforce86** [reply]
Ah okay thank you! Once I'm back at my desk I'll give that a try.

FWIW I am already including it that way as well, but I assume one must be overwriting parts of the other?

**Reply by bassforce86** [reply]
Replying on mobile is apparently delayed 😂.

Thank you for getting back to me so swiftly! It's much appreciated

**Reply by bassforce86** [reply]
The plot thickens... So, removing the `nixvim` module did indeed solve my problem _but only on my ChromeOS & Linux_ environments. 

But not in my darwin environment... Is there perhaps some path linking that's different with regards to Neovim's config somewhere for Mac?

**Reply by bassforce86** [reply]
(for debugging purposes, you can ignore ChromeOS - it's essentially a debian VM)

**Reply by bassforce86** [reply]
Interestingly - I lied, on a second (unrelated) switch it's also still broken on linux and chromeos 

running `:checkhealth` has a warning about no init.lua config found... but I _assume_ that's because it's byte compiled? (as per config/performance.nix)

**Reply by MattSturgeon** [reply]
> running `:checkhealth` has a warning about no init.lua config found...

That happens because we call nvim with the `-u` flag to specify a custom init file. We also remove `~/.config.nvim` from the runtimepath to avoid potential impurity.

If you're getting that error, it sounds like it is installed correctly, but maybe something in your generated config is broken?

Does `nixvim-print-init` have useful output?

Also note: some of the performance options are experimental and can lead to broken configs in some scenarios.

**Reply by bassforce86** [reply]
From what I can tell, the only thing `nixvim-print-init` tells me is that dashboard is loaded very late in the plugin order, so netrw is loading and not allowing dashboard to show. It might just be a dashboard thing potentially? 

if I go to a file, then run `:Dashboard` i can bring up the dashboard, but that semi-defeats the purpose i intended to use the dashboard for (which is a jumping-off point).

Happy to copy the whole output from `nixvim-print-init` if it'll help?

**Reply by bassforce86** [reply]
It's also possible this is a byte compilation / performance issue - although I have also tried running this with performance disabled and it didn't appear to have an affect (but i have also made numerous changes since then too)

**Reply by MattSturgeon** [reply]
It sounds like it may be another import-order issue brought on by https://github.com/nix-community/nixvim/pull/2164.

In your neovim flake, try overriding to an older revision and running with `nix run` or `nix shell`.

```
nix flake lock --override-input nixvim github:nix-community/nixvim/faff32b9f1f5ada447d2ddfb4dcdcbeb7ec01c0b

nix shell

nixvim-print-init

vim
```

**Reply by bassforce86** [reply]
Sorry for the delayed reply, `nix run` works as expected with the dashboard showing. Will try with `nix shell` and see if that produces the same result.

**Reply by bassforce86** [reply]
Out of curiosity, did you mean `nix shell`? would working with flakes not usually be `nix develop`? (i'm honestly unsure which takes precedence these days)

**Reply by MattSturgeon** [reply]
`nix shell` and `nix develop` do different things.

`nix shell` enters a shell where the package is available on the PATH, similar to installing the package.

`nix shell` followed by running the main program is effectively equivalent to `nix run`.

`nix develop` enters a shell derivation explicitly defined in the flake outputs. Usually this'd be something defined using `pkgs.mkShell`. It can also be used to enter a shell where a package's build dependencies are all available for use.

TL;DR `nix shell` and `nix run` are for using a package, `nix develop` is (usually) for building or debugging a package.

**Reply by MattSturgeon** [reply]
Could you clarify though, did you need to do the suggested `flake lock` command, or does running with `nix run` work even without modifying your lock file?

If so, it is possible you're just forgetting to update your downstream config's lockfile when you change your upstream nixvim config.

**Reply by bassforce86** [reply]
Sure, `nix run` works regardless of the nix lock downgrade. but it never works standalone - I have ensured I have updated the lock in home-manager, but no joy - dashboard still doesn't load. - I've pushed the updates to my home-manager flake in case anybody has the time to pull and investigate.

**Reply by bassforce86** [reply]
I've also tried with the patch to the home-manager flake to ensure the downgrade also applied when running through standalone / home-manager - but still no change. No dashboard sadly.

**Reply by MattSturgeon** [reply]
Forgive me for ruling out the obvious things first, but if your nixvim config works when run directly from its flake then the config is good.

I think either you're having workflow issues keeping the two configs in sync with each other, or maybe you're forgetting to build & activate your home-manager config after updating/syncing it?

If hm is set up as a nixos modules, you'd do this via `sudo nixos-rebuild switch --flake .`. If it is setup standalone then you'd use `home-manager switch --flake .`.

As for your lockfile sync, you should ensure that first you're pushing your changes on your nixvim flake, then use `nix flake update` (for everything) or `nix flake lock --update-input <input-name>` (for just your neovim flake) to sync up your main config, verify the hash in your lockfile matches what you pushed on your nixvim flake, then build & activate your new config.

**Reply by bassforce86** [reply]
> Forgive me for ruling out the obvious things first, but if your nixvim config works when run directly from its flake then the config is good.

Not at all - that's all part of triage, I am human, I make (many) mistakes 😄 . I came to a similar conclusion - hence the request for helping with debugging. The non-deterministic output confused me.

I use `home-manager switch --flake .` after a `nix flake update` in the home-manager repo, but still don't get the dashboard displayed when opening neovim via `nvim, vim, or vi`

**Reply by MattSturgeon** [reply]
Does `nixvim-print-init` after activating your hm config print the same thing as `nixvim-print-init` run within `nix shell` on your nixvim flake?

**Reply by bassforce86** [reply]
just to confirm i am making the correct calls to give you accurate information, running `nix shell .` in the neovim flake dir, is what you mean by checking the output in nix-shell?

Below is a side-by-side of `nixvim-print-init` from `nix shell .` (left) and `nixvim-print-init` from current home-manager generation. 
Strangely, they are both referencing the same store reference... Which is not what I would expect when running something from within nix shell.

<img width="1764" alt="Screenshot 2024-09-13 at 11 19 34" src="https://github.com/user-attachments/assets/803c048d-8535-42f6-b3b7-fddc9be11be0">

**Reply by MattSturgeon** [reply]
The files being the same in the nix-store _is_ expected when both are built from the same config, since nix tries to be pure and deterministic.

The _symlinks to_ the files will be different, since they are made available to your shell differently depending on how you activate/install them.

However something very strange is going on if you have the same init.lua file in both environments, but one of them works and the other does not.

Perhaps it is something impure from outside your nixvim config having an effect at runtime? At this point I'm just taking wild guesses though :confounded:

**Reply by bassforce86** [reply, marked as answer]
So, I've no idea why the non-deterministic-ness is occuring, but i've decided to work around the problem for now.

For anyone curious, here's the relevant autocmd: https://gitlab.com/bf86/flakes/neovim/-/blob/main/config/autocmd.nix?ref_type=heads#L81

And here, i've skipped loading netrw on startup just incase: https://gitlab.com/bf86/flakes/neovim/-/blob/main/config/globals.nix?ref_type=heads#L12

This now works as expected.
